### DISCUSSION TITLE
How to calculate a windowed rank function?

### DISCUSSION THREAD

**Discussion by daviewales** [saulpw/visidata#2280] [Q&A] [answered]
Suppose I have some sorted data like this:

| test |
| ---- |
| A    |
| A    |
| B    |
| C    |
| C    |
| C    |
| D    |
| D    |

And I want to rank the items in order, but restart the rank each time the group changes:

| test | rank |
| ---- | ---- |
| A    |    1 |
| A    |    2 |
| B    |    1 |
| C    |    1 |
| C    |    2 |
| C    |    3 |
| D    |    1 |
| D    |    2 |

Can I do this with Visidata?

The closest I've gotten is using the Window function: `w` with the argument `1 0`, then creating a new =Python column:

``` python
=1 if test != test_previous else None
```

![image](https://github.com/saulpw/visidata/assets/2499066/7168eee4-a14d-4355-bdb8-cfdbd1a84bcd)

This gives me the first value of each group, but doesn't give me the subsequent ranks.
I saw this issue which suggests it might not currently be possible, but thought I'd ask in case there was another way: https://github.com/saulpw/visidata/issues/1399

(Note, also working around #2279 by creating a new frozen column with `'` and manually setting the value of the first cell to NULL with `zd`.)

**Comment by midichef** [comment, marked as answer]
Hi, yes, you can do this with Visidata. It took me a while to remember how. I did it with your method: mark the start of the group. The other key steps are:
1) Get the row number, using `i` (`addcol-incr`). Rename that column to `i`:  `^` `i`
2) Let each row see the row number of the start of its group:
a) Make a window with `addcol-window` to use the value of the previous row and the value on the current row:  `w` `1 0`
b) Then, make an expression coumn that will be called `first`.  The expression to use [has changed as of September 4, 2024](https://github.com/saulpw/visidata/commit/bbfbb45eec68f7ee8745008f1428cc23565dae88).
For older visidata versions than that, like visidata <= 3.0.2:
    i) Make an expression column which will be called `first`:  `=` `w[0] != w[1] if len(w) == 2 else True` and rename it:  `^` `first`.
For newer versions >= v3.1dev, with commits after September 4, 2024, instead change the expression:
    ii) Make an expression column which will be called `first`:  `=` `w[0] != w[1] if i > 1 else True` and rename it:  `^` `first`.
c) Make a new expression column `=` `i if first else None`. Select all rows with `gs`, then `f` to fill nulls. Rename it: `^` `filled`
d) Then make a final column with the ranks: `=` `i-filled+1`.
![rank](https://github.com/saulpw/visidata/assets/67946319/2e8441de-92f5-431e-a969-31a92d5aeb10)

**Reply by daviewales** [reply]
That is extremely clever. A bit too clever for me to remember it I'm sure, but it will be here for future reference now. Thanks!

**Comment by midichef** [comment]
There is a much simpler way to rank within groups, since visidata version 3.2.
1. press `!` on the column you want to group by for ranking, since groups are determined by key columns
2. `addcol-incr` (`i`) to add values that resolve ties within a group.
3. while on that new column of numbers, `addcol-aggregate` `rank`.
